_posts/2019-01-03-elixir-kaffe-codealong.md

```diff
@@ -0,0 +1,204 @@
+```elixir
+def application do
+  [
+    extra_applications: [:logger, :kaffe],
+    mod: {ElixirKaffeCodealong.Application, []}
+    # now that we're using the Application module, this is where we'll tell it to start.
+    # We use the keyword `mod` with applications that start a supervision tree,
+    # which we configured when adding our Kaffe.Consumer to Application above.
+  ]
+end
+```
+**4.c:** add a consumer module to accept messages from Kafka as `/lib/example_consumer.ex` with the following code:
+```elixir
+defmodule ExampleConsumer do
+  # function to accept Kafka messaged MUST be named "handle_message"
+  # MUST accept arugments structured as shown here
+  # MUST return :ok
+  # Can do anything else within the function with the incoming message
+
+  def handle_message(%{key: key, value: value} = message) do # necessary
+    IO.inspect message # optional
+    IO.puts "#{key}: #{value}" # optional
+    :ok # necessary
+  end
+end
+```
+**4.d:** configure the consumer module in `/config/config.exs`
+```elixir
+config :kaffe,
+  consumer: [
+    endpoints: [localhost: 9092],               
+    topics: ["our_topic", "another_topic"],     # the topic(s) that will be consumed
+    consumer_group: "example-consumer-group",   # the consumer group for tracking offsets in Kafka
+    message_handler: ExampleConsumer,           # the module that will process messages
+  ]
+```
+
+**5. Add a producer module (optional, can also call Kaffe from the console)**
+add `/lib/example_producer.ex` with the following code:
+```elixir
+defmodule ExampleProducer do
+  def send_my_message({key, value}, topic) do
+    Kaffe.Producer.produce_sync(topic, [{key, value}])
+  end
+
+  def send_my_message(key, value) do
+    Kaffe.Producer.produce_sync(key, value)
+  end
+
+  def send_my_message(value) do
+    Kaffe.Producer.produce_sync("sample_key", value)
+  end
+end
+```
+
+**6. Send and receive messages in the console!**
+Start an interactive elixir shell with `iex -S mix` and call the following:
+```sh
+iex> ExampleProducer.send_my_message({"Metamorphosis", "Franz Kafka"}, "another_topic")
+...>[debug] event#produce_list topic=another_topic
+...>[debug] event#produce_list_to_topic topic=another_topic partition=0
+...>:ok
+iex> %{
+...> attributes: 0,
+...> crc: 2125760860, # will vary
+...> key: "Metamorphosis",
+...> magic_byte: 1,
+...> offset: 1, # will vary
+...> partition: 0,
+...> topic: "another_topic",
+...> ts: 1546634470702, # will vary
+...> ts_type: :create,
+...> value: "Franz Kafka"
+...> }
+...> Metamorphosis: Franz Kafka
+```
+
+## Variations: Docker & Umbrella Apps
+- If you're running Kafka from a docker container (most common in real applications), you will use that hostname in the config file rather than `localhost`
+- In an umbrella app you'll configure Kaffe in the child application running it. If you have apps separated by environment, you can start the consumer by structuring it as a child like this:
+```elixir
```

Comment:
I think I don't totally get this example. You would have `ElixirKaffeCodealong.OtherApp` running in dev but not production?

Reply:
I will add it to the production one too - that was my oversight. You can definitely have it do this but you're totally right that it would be weird. You may not want it running in your test env, so all I was showing here is that you can configure which environments they're running in. But for this example I'll change it to be used in dev and prod.